### PR TITLE
Perf:  fix restore performance regression

### PR DESCRIPTION
**Pull request by dtivel**
Improvement over #1545.

This is for internal bug 451335.

### CODE REVIEW

**Comment by alpaix** [comment]
Is this final improvement? What would be the impact on low memory boxes? Can you test the perf on x86 Windows?

File: src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs [hunk 1]
```diff
@@ -144,23 +144,20 @@ public void Dispose()
...
             using (var destination = new FileStream(
                 destinationFilePath,
                 FileMode.Create,
-                FileAccess.ReadWrite,
-                FileShare.ReadWrite | FileShare.Delete,
-                bufferSize: 4096,
+                FileAccess.Write,
+                FileShare.Write,
+                bufferSize,
                 useAsync: true))
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs]: do you see better perf when setting useAsync to false? I recall a perf issue in the past where using async caused additional security checks each time a different thread resumed the operation, and the recommendation was to just not use async file streams. @alpaix would know better.

File: src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs [hunk 2]
```diff
@@ -144,23 +144,20 @@ public void Dispose()
...
+            // This value is copied from System.IO.Stream.
+            // It is the largest multiple of 4096 that is smaller than the large object heap (LOH)
+            // threshold of 85K and offers better copy performance than a 4K buffer.
+            const int bufferSize = 81920;
+
             using (var destination = new FileStream(
                 destinationFilePath,
                 FileMode.Create,
-                FileAccess.ReadWrite,
-                FileShare.ReadWrite | FileShare.Delete,
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs]: why did we change the fileaccess and fileshare? does it also impact perf?

File: src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs (hunk 1 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs]: Discussed offline, going by the overall perf numbers is best here.

**Review by emgarten — APPROVED** [review]

**Comment by dtivel** [comment]
@alpaix, yes, this is the final improvement.  With the latest commit, there are no memory concerns from the baseline.

File: src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs (hunk 2 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs]: @jainaashish, I'll make this change at a later time.  It's not perf-related.  This method copies a .nupkg and read access and read/delete sharing is unnecesssary.  This code dates from before plugin changes, and at that time the same stream was used for writing, reading and deleting.  In a post-plugin world, that code has been split up.

File: src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs [hunk 3]
```diff
@@ -144,23 +144,21 @@ public void Dispose()
 
             cancellationToken.ThrowIfCancellationRequested();
 
+            using (var source = File.OpenRead(_packageFilePath))
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs:147]: how does this improve the original perf issue?

Comment (dtivel) [reply src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs:147]: It uses `useAsync` `false`.  Previously, the source was `_sourceStream.Value` which was created with `useAsync` `true`.

Comment (jainaashish) [reply src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs:147]: so if it isn't `async` anymore should we be really using `CopyToAsync`?

Comment (dtivel) [reply src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs:147]: In my restore perf testing `CopyTo(...)` is roughly twice as slow as `CopyToAsync(...)` with all other factors the same.

**Review by jainaashish — APPROVED** [review]

**Pull request merged**